kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/DomainMetadata.java

```diff
@@ -0,0 +1,124 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.actions;
+
+import static io.delta.kernel.internal.util.InternalUtils.requireNonNull;
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.data.GenericRow;
+import io.delta.kernel.types.BooleanType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.types.StructType;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Delta log action representing an `DomainMetadata` action */
+public class DomainMetadata {
+
+  public static DomainMetadata fromColumnVector(ColumnVector vector, int rowId) {
+    if (vector.isNullAt(rowId)) {
+      return null;
+    }
+
+    return new DomainMetadata(
+        requireNonNull(vector.getChild(0), rowId, "domain").getString(rowId),
+        requireNonNull(vector.getChild(1), rowId, "configuration").getString(rowId),
+        requireNonNull(vector.getChild(2), rowId, "removed").getBoolean(rowId));
+  }
+
+  public static DomainMetadata fromRow(Row row) {
+    assert (row.getSchema() == FULL_SCHEMA);
+    return new DomainMetadata(
+        requireNonNull(row, 0, "domain").getString(0),
+        requireNonNull(row, 1, "configuration").getString(1),
+        requireNonNull(row, 2, "removed").getBoolean(2));
+  }
+
+  /** Full schema of the {@link DomainMetadata} action in the Delta Log. */
+  public static final StructType FULL_SCHEMA =
+      new StructType()
+          .add("domain", StringType.STRING, false /* nullable */)
+          .add("configuration", StringType.STRING, false /* nullable */)
+          .add("removed", BooleanType.BOOLEAN, false /* nullable */);
+
+  /** The feature name for domain metadata. */
+  public static final String FEATURE_NAME = "domainMetadata";
+
+  /** The minimum writer version required to support domain metadata. */
+  public static final int MIN_WRITER_VERSION_REQUIRED = 7;
+
+  private final String domain;
+  private final String configuration;
+  private final boolean removed;
+
+  /**
+   * The domain metadata action contains a configuration (string-string map) for a named metadata
+   * domain. Two overlapping transactions conflict if they both contain a domain metadata action for
+   * the same metadata domain. Per-domain conflict resolution logic can be implemented.
+   *
+   * @param domain A string used to identify a specific feature.
+   * @param configuration A string containing configuration options for the conflict domain.
+   * @param removed If it is true it serves as a tombstone to logically delete a {@link
+   *     DomainMetadata} action.
+   */
+  public DomainMetadata(String domain, String configuration, boolean removed) {
```

Comment:
Do you suggest we should also parse and validate the `configuration` string as a `Map<String, String>` at construction, in addition to storing a raw string? Currently I only use the raw string since this is how it is write to/read from log files and this is how it is implemented in DBR.

Reply:
1. On read path, it can be lazy.
2. On write path, we should be sure we only write valid domain metadata. Btw, where in the spec does it say the configuration has to be a `Map<String, String>` ?
3. Shouldn't we validate on the write path that the user hasn't written to a `System-controlled metadata domain`?

Reply:
> "The domain metadata action contains a configuration (string-string map) for a named metadata domain"

Ah I'm sorry - my comment in the code^ was incorrect and caused some confusion. I’ve removed the `string-string map` from the description. Let me clarify my thinking here.

The Delta Protocol only requires the configuration to be a string containing configuration data, without enforcing a specific format. So I think it could technically be in any format and doesn't have to be a `Map<String, String>`.

In practice, what we did in DBR is, in addition to the `DomainMetadata` action class, we have a `JsonMetadataDomain` trait particularly used for domains where the configuration string is in JSON format. It also has utility functions to serialize/deserialize the domain object into a JSON string. I checked DBR and it seems all metadata domains we are currently using extends `JsonMetadataDomain`. Even though JSON is commonly used, we probably shouldn’t require it strictly for the configuration string.

In Kernel, I think we could do the same: we could have a general `DomainMetadata` class to represent the action, allowing any string for the configuration, and then implement a similar `JsonMetadataDomain` for cases where JSON configuration string is required, with necessary validations/serialization/deserialization functions for it (will be introduced in a later PR). Does this make sense to you? 



> A tangential question: how is domain metadata actually used? What do internal code paths do with the configuration string?

In DBR, we have multiple feature metadata domain (e.g. `RowTrackingMetadataDomain`, `ClusteringMetadataDomain`), which all extend `JsonMetadataDomain`, with their own domain-specific configuration fields .

For example, if we need access to the `RowTrackingMetadataDomain.rowIdHighWaterMark`, we first find the `DomainMetadata` by comparing the domain name, and then deserialize the JSON configuration string back into the typed domain object (`RowTrackingMetadataDomain`) to access the field. And to create a `DomainMetadata` action for commit, we can call `toDomainMetadata` on a `RowTrackingMetadataDomain` object to serialize domain-specific configuration fields to JSON and create the `DomainMetadata` action.

Please let me know if anything is unclear. Thanks!

Reply:
Makes sense. Thanks for the explanation!